src/coreclr/System.Private.CoreLib/src/System/Variant.cs

```diff
@@ -78,323 +23,296 @@ internal partial struct Variant
-                    TypeCode.Double => new Variant(ic.ToDouble(provider)),
-                    TypeCode.Decimal => new Variant(ic.ToDecimal(provider)),
-                    TypeCode.DateTime => new Variant(ic.ToDateTime(provider)),
-                    TypeCode.String => new Variant(ic.ToString(provider)),
-                    _ => throw new NotSupportedException(SR.Format(SR.NotSupported_UnknownTypeCode, ic.GetTypeCode())),
-                };
-            }
-            else
+            // Invalid and common types are handled at native side
+            Debug.Assert((pOle.VarType & VarEnum.VT_ARRAY) == 0, "Array should be handled at native side.");
+            Debug.Assert((pOle.VarType & ~VarEnum.VT_BYREF) != VarEnum.VT_RECORD, "Records should be handled at native side.");
+
+            switch (pOle.VarType)
             {
-                // This path should eventually go away. But until
-                // the work is done to have all of our wrapper types implement
-                // IConvertible, this is a cheapo way to get the work done.
-                v = new Variant(o);
+                case VarEnum.VT_I4:
+                case VarEnum.VT_INT:
+                    return pOle.As<int>();
+                case VarEnum.VT_BYREF | VarEnum.VT_I4:
+                case VarEnum.VT_BYREF | VarEnum.VT_INT:
+                    return *(int*)pOle.GetRawDataRef<IntPtr>();
+
+                case VarEnum.VT_UI4:
+                case VarEnum.VT_UINT:
+                    return pOle.As<uint>();
+                case VarEnum.VT_BYREF | VarEnum.VT_UI4:
+                case VarEnum.VT_BYREF | VarEnum.VT_UINT:
+                    return *(uint*)pOle.GetRawDataRef<IntPtr>();
+
+                case VarEnum.VT_I1:
+                    return pOle.As<sbyte>();
+                case VarEnum.VT_BYREF | VarEnum.VT_I1:
+                    return *(sbyte*)pOle.GetRawDataRef<IntPtr>();
+
+                case VarEnum.VT_UI1:
+                    return pOle.As<byte>();
+                case VarEnum.VT_BYREF | VarEnum.VT_UI1:
+                    return *(byte*)pOle.GetRawDataRef<IntPtr>();
+
+                case VarEnum.VT_I2:
+                    return pOle.As<short>();
+                case VarEnum.VT_BYREF | VarEnum.VT_I2:
+                    return *(short*)pOle.GetRawDataRef<IntPtr>();
+
+                case VarEnum.VT_UI2:
+                    return pOle.As<ushort>();
+                case VarEnum.VT_BYREF | VarEnum.VT_UI2:
+                    return *(ushort*)pOle.GetRawDataRef<IntPtr>();
+
+                case VarEnum.VT_I8:
+                    return pOle.As<long>();
+                case VarEnum.VT_BYREF | VarEnum.VT_I8:
+                    return *(long*)pOle.GetRawDataRef<IntPtr>();
+
+                case VarEnum.VT_UI8:
+                    return pOle.As<ulong>();
+                case VarEnum.VT_BYREF | VarEnum.VT_UI8:
+                    return *(ulong*)pOle.GetRawDataRef<IntPtr>();
```

Comment:
It isn't uncommon, but we try to avoid adding new features to these code paths since they are observable - it is common for users to rely on exceptions for control flow.

Reply:
It's not new feature. In existing code, it's not handled at fast path at native side, and goes through slow path that involves managed code.

The questions is whether we should bring the optimization.

Reply:
> The questions is whether we should bring the optimization.

Yes.